woocommerce/woocommerce-gateway-stripe
# Use deferred intents on the My Account > Add Payment Method page

Fixes #2762 

This PR updates the **My Account > Add Payment Method** flow to use deferred intents.

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/781e4f0a-07d2-49e7-9bb5-19d6892a2e63)

**Changes proposed in this Pull Request:**

 - 164db56d85031f9bbd8be49d3da586dd14dccb1f introduces a new `api.setupIntent()` function which is called by `createAndConfirmSetupIntent()`. This function sends off an AJAX request to create and confirm a setup intent using the new payment method ID (pm_)
 - https://github.com/woocommerce/woocommerce-gateway-stripe/commit/3073941839651c8466b7da1301f5775e718fcd07 adds a new `createAndConfirmSetupIntent()` function which gets passed to the `processPayment()` as the `additionalActionsHandler` arg when the `form#add_payment_method` is submitted. This function calls the new setupIntent function above and adds the setup intent ID to a hidden form before submitting the form.
 - https://github.com/woocommerce/woocommerce-gateway-stripe/commit/34b0b310a75cc517210fb502b80b6d3797d16acb removes the old and unused code that mounted the previous Stripe UPE and adds code to load the new Stripe PE on the Add payment method page.
 - https://github.com/woocommerce/woocommerce-gateway-stripe/commit/1a16c6c3d33c4dc7fca5473467980908703e7b0f Introduces backend code to handle the AJAX requests to create and confirm a setup intent with the new payment method ID.
 - https://github.com/woocommerce/woocommerce-gateway-stripe/commit/375b651daca16f8b0f2838e9277d1442da9bd618 Handles the submission of the Add payment method form by creating a new token stored in the `wp_woocommerce_payment_tokens` DB table.
 - https://github.com/woocommerce/woocommerce-gateway-stripe/commit/809541992c069fbc51a456726f06a0afcd4f8a0a This commit fixes an issue I was having when trying to load the Stripe PE with all of the payment method types. Some payment method types don't support setup intents and $0 amounts so I changed this to just return 'card' types. As we implement Split PE, we can look into this.



## Testing instructions





1. Open up the Stripe events page in your dashboard: https://dashboard.stripe.com/test/events
3. While on `develop`, go to My Account > Payment method > Add payment method
4. When the form loads, visit the Stripe events page and notice the `setup_intent.created` is triggered.
5. Checkout this branch, then run `npm run build:webpack`
6. Refresh the Add payment method page and note that there's no `setup_intent.created` happening on page load.
7. Continue to add a new card `4242424242424242` and submit the form.
8. Confirm you are redirected to the My Account > Payment methods page and your new card is added
9. Delete the card from this page and confirm it's been removed.
10. Go back to the Add payment method page and attempt to add a bad card `4000000000000002` and you are presented with an error notice.

---


**Post merge**
